File: artifacts/api-server/src/py-api/routes/userReviews.ts
import { Router } from "express";
import { query, queryOne, execute, PREDEFINED_USERS } from "../db.js";
import { v4 as uuidv4 } from "uuid";
import { runReviewAgent } from "../agents/reviewAgent.js";
import { getAllDetectorOpinions, getDetectorOpinion, DETECTOR_AGENTS } from "../agents/detectorAgents.js";

const router = Router();

// ── 1. Profiles / User list ──────────────────────────────────────────────────
router.get("/users", async (_req, res) => {
  try {
    const dbUsers = await query("SELECT id, username, display_name, created_at FROM dup_users");
    const byId = new Map(dbUsers.map((u) => [u["id"], u]));
    const users = PREDEFINED_USERS.map((u) => ({
      id: u.id,
      username: u.username,
      display_name: u.displayName,
      displayName: u.displayName,
      createdAt: byId.get(u.id)?.["created_at"] ?? new Date().toISOString(),
    }));
    res.json({ users });
  } catch (e) {
    console.error("Get users error:", e);
    res.status(500).json({ error: "Failed to get users" });
  }
});

// ── 2. Fetch (assign) queue ───────────────────────────────────────────────────
router.post("/fetch", async (req, res) => {
  try {
    const { userId, limit = 20 } = req.body as { userId: string; limit?: number };
    const user = PREDEFINED_USERS.find((u) => u.id === userId);
    if (!user) return res.status(404).json({ error: "User not found" });

    const fetchLimit = Math.min(100, Math.max(1, Number(limit)));
    const pending = await query(
      `SELECT id FROM dup_duplicate_payments WHERE status = 'pending' OR status = 'under_review' ORDER BY probability DESC LIMIT $1`,
      [fetchLimit * 3]
    );

    const prevTotal = await query<{ total: string }>(
      "SELECT COUNT(*) AS total FROM dup_user_reviews WHERE user_id = $1",
      [userId]
    );
    const prevTotalCount = Number(prevTotal[0]?.total || 0);

    let newlyAssigned = 0;
    for (const p of pending) {
      const exists = await queryOne(
        "SELECT id FROM dup_user_reviews WHERE user_id = $1 AND duplicate_payment_id = $2",
        [userId, p["id"]]
      );
      if (!exists) {
        await execute(
          `INSERT INTO dup_user_reviews (id, user_id, duplicate_payment_id, status, created_at, updated_at) VALUES ($1, $2, $3, 'pending', NOW(), NOW())`,
          [uuidv4(), userId, p["id"]]
        );
        newlyAssigned++;
        if (newlyAssigned >= fetchLimit) break;
      }
    }

    const newTotal = await query<{ total: string }>(
      "SELECT COUNT(*) AS total FROM dup_user_reviews WHERE user_id = $1",
      [userId]
    );
    const totalAssigned = Number(newTotal[0]?.total || 0);

    return res.json({
      totalFetched: newlyAssigned,
      newlyAssigned,
      totalAssigned,
      summary: newlyAssigned > 0
        ? `Assigned ${newlyAssigned} new duplicate payment${newlyAssigned !== 1 ? "s" : ""} to ${user.displayName}'s review queue. Total queue size: ${totalAssigned}.`
        : `No new payments to assign. ${user.displayName} already has ${totalAssigned} payment${totalAssigned !== 1 ? "s" : ""} in their queue.`,
    });
  } catch (e) {
    console.error("Fetch reviews error:", e);
    return res.status(500).json({ error: "Failed to fetch reviews" });
  }
});

// ── 3. List reviews for a user ───────────────────────────────────────────────
router.get("/:userId", async (req, res) => {
  try {
    const { userId } = req.params;
    const page = Math.max(1, Number(req.query["page"] || 1));
    const pageSize = Math.min(100, Math.max(1, Number(req.query["page_size"] || 20)));
    const offset = (page - 1) * pageSize;
    const statusFilter = req.query["status"] as string | undefined;

    const params: unknown[] = [userId];
    let where = "WHERE ur.user_id = $1";
    if (statusFilter) { where += " AND ur.status = $2"; params.push(statusFilter); }

    const [countRow] = await query<{ total: string }>(
      `SELECT COUNT(*) AS total FROM dup_user_reviews ur ${where}`, params
    );
    const total = Number(countRow.total);

    const reviews = await query(
      `SELECT ur.id, ur.user_id, ur.duplicate_payment_id, ur.status, ur.notes, ur.created_at, ur.updated_at,
              dp.payment1_id, dp.payment2_id, dp.probability, dp.duplicate_type, dp.payment_system,
              dp.amount, dp.currency, dp.sender_bic, dp.receiver_bic, dp.originator_country,
              dp.beneficiary_country, dp.payment_date1, dp.payment_date2, dp.matched_fields, dp.detected_at
       FROM dup_user_reviews ur
       JOIN dup_duplicate_payments dp ON dp.id = ur.duplicate_payment_id
       ${where}
       ORDER BY dp.probability DESC, ur.created_at DESC
       LIMIT $${params.length + 1} OFFSET $${params.length + 2}`,
      [...params, pageSize, offset]
    );

    res.json({
      total,
      page,
      pageSize,
      pages: Math.max(1, Math.ceil(total / pageSize)),
      reviews: reviews.map((r) => ({
        id: r["id"],
        userId: r["user_id"],
        duplicatePaymentId: r["duplicate_payment_id"],
        status: r["status"],
        notes: r["notes"],
        createdAt: r["created_at"],
        updatedAt: r["updated_at"],
        payment: {
          id: r["duplicate_payment_id"],
          payment1Id: r["payment1_id"],
          payment2Id: r["payment2_id"],
          probability: Number(r["probability"] || 0),
          duplicateType: r["duplicate_type"],
          paymentSystem: r["payment_system"],
          amount: Number(r["amount"] || 0),
          currency: r["currency"],
          senderBIC: r["sender_bic"],
          receiverBIC: r["receiver_bic"],
          originatorCountry: r["originator_country"],
          beneficiaryCountry: r["beneficiary_country"],
          paymentDate1: r["payment_date1"],
          paymentDate2: r["payment_date2"],
          matchedFields: r["matched_fields"],
          detectedAt: r["detected_at"],
        },
      })),
    });
  } catch (e) {
    console.error("List reviews error:", e);
    res.status(500).json({ error: "Failed to list reviews" });
  }
});

// ── 4. Get chat messages for a review ─────────────────────────────────────────
router.get("/:userId/:reviewId/messages", async (req, res) => {
  try {
    const { userId, reviewId } = req.params;
    const review = await queryOne(
      `SELECT ur.status FROM dup_user_reviews ur WHERE ur.id = $1 AND ur.user_id = $2`,
      [reviewId, userId]
    );
    if (!review) return res.status(404).json({ error: "Review not found" });

    const rows = await query(
      "SELECT * FROM dup_user_review_messages WHERE review_id = $1 ORDER BY timestamp ASC",
      [reviewId]
    );
    return res.json({
      messages: rows.map((r) => ({ id: r["id"], role: r["role"], content: r["content"], timestamp: r["timestamp"] })),
      currentStatus: review["status"],
    });
  } catch (e) {
    console.error("Get review messages error:", e);
    return res.status(500).json({ error: "Failed to get messages" });
  }
});

// ── 5. Detector opinions for a review ─────────────────────────────────────────
router.get("/:userId/:reviewId/opinions", async (req, res) => {
  try {
    const { userId, reviewId } = req.params;
    const review = await queryOne(
      `SELECT ur.duplicate_payment_id FROM dup_user_reviews ur WHERE ur.id = $1 AND ur.user_id = $2`,
      [reviewId, userId]
    );
    if (!review) return res.status(404).json({ error: "Review not found" });

    const dup = await queryOne("SELECT * FROM dup_duplicate_payments WHERE id = $1", [review["duplicate_payment_id"]]);
    if (!dup) return res.status(404).json({ error: "Duplicate payment not found" });

    const memoryRows = await query("SELECT content FROM dup_agent_memory ORDER BY updated_at DESC LIMIT 20");
    const memoryContext = memoryRows.map((r) => r["content"]).join("\n\n");

    const { opinions, consensus } = await getAllDetectorOpinions([dup], memoryContext);
    return res.json({ opinions, consensus });
  } catch (e) {
    console.error("Detector opinions error:", e);
    return res.status(500).json({ error: "Failed to get detector opinions" });
  }
});

// ── 6. Chat (Training Agent per review) ───────────────────────────────────────
router.post("/:userId/:reviewId/chat", async (req, res) => {
  try {
    const { userId, reviewId } = req.params;
    const { message } = req.body as { message: string };

    const user = PREDEFINED_USERS.find((u) => u.id === userId);
    if (!user) return res.status(404).json({ error: "User not found" });

    const review = await queryOne(
      `SELECT ur.*, dp.payment1_id, dp.payment2_id, dp.probability, dp.duplicate_type,
              dp.payment_system, dp.amount, dp.currency, dp.sender_bic, dp.receiver_bic,
              dp.originator_country, dp.beneficiary_country, dp.payment_date1, dp.payment_date2,
              dp.matched_fields
       FROM dup_user_reviews ur
       JOIN dup_duplicate_payments dp ON dp.id = ur.duplicate_payment_id
       WHERE ur.id = $1 AND ur.user_id = $2`,
      [reviewId, userId]
    );
    if (!review) return res.status(404).json({ error: "Review not found" });

    const paymentData = {
      id: review["duplicate_payment_id"],
      payment1Id: review["payment1_id"],
      payment2Id: review["payment2_id"],
      probability: review["probability"],
      duplicateType: review["duplicate_type"],
      paymentSystem: review["payment_system"],
      amount: review["amount"],
      currency: review["currency"],
      senderBIC: review["sender_bic"],
      receiverBIC: review["receiver_bic"],
      originatorCountry: review["originator_country"],
      beneficiaryCountry: review["beneficiary_country"],
      paymentDate1: review["payment_date1"],
      paymentDate2: review["payment_date2"],
      matchedFields: review["matched_fields"],
      status: review["status"],
    };

    const savedMsgId = uuidv4();
    await execute(
      `INSERT INTO dup_user_review_messages (id, review_id, role, content, timestamp) VALUES ($1, $2, 'user', $3, NOW())`,
      [savedMsgId, reviewId, message]
    );

    const dbHistory = await query(
      "SELECT role, content FROM dup_user_review_messages WHERE review_id = $1 ORDER BY timestamp ASC LIMIT 100",
      [reviewId]
    ) as Array<{ role: string; content: string }>;

    const { response, statusUpdate } = await runReviewAgent({
      userMessage: message,
      userId,
      reviewerName: user.displayName,
      paymentData,
      detectorOpinions: [],
      dbHistory,
    });

    const replyId = uuidv4();
    await execute(
      `INSERT INTO dup_user_review_messages (id, review_id, role, content, timestamp) VALUES ($1, $2, 'assistant', $3, NOW())`,
      [replyId, reviewId, response]
    );

    if (statusUpdate) {
      await execute(
        `UPDATE dup_user_reviews SET status = $1, updated_at = NOW() WHERE id = $2`,
        [statusUpdate, reviewId]
      );
      await syncGlobalStatus(review["duplicate_payment_id"] as string);
    }

    const currentReview = await queryOne("SELECT status FROM dup_user_reviews WHERE id = $1", [reviewId]);

    return res.json({
      userMessageId: savedMsgId,
      response,
      assistantMessageId: replyId,
      statusUpdate,
      currentStatus: currentReview?.["status"] ?? review["status"],
    });
  } catch (e) {
    console.error("Review chat error:", e);
    return res.status(500).json({ error: "Chat failed" });
  }
});

// ── 7. Update status for a review (PATCH /:userId/:reviewId/status) ──────────
router.patch("/:userId/:reviewId/status", async (req, res) => {
  try {
    const { userId, reviewId } = req.params;
    const { status, notes } = req.body as { status: string; notes?: string };
    const validStatuses = ["pending", "confirmed_duplicate", "dismissed", "under_review"];
    if (!validStatuses.includes(status)) return res.status(400).json({ error: "Invalid status" });

    const params: unknown[] = [status, new Date().toISOString()];
    let sql = `UPDATE dup_user_reviews SET status = $1, updated_at = $2`;
    if (notes !== undefined) { sql += `, notes = $${params.length + 1}`; params.push(notes); }
    sql += ` WHERE id = $${params.length + 1} AND user_id = $${params.length + 2} RETURNING *`;
    params.push(reviewId, userId);

    const rows = await query(sql, params);
    if (!rows.length) return res.status(404).json({ error: "Review not found" });

    await syncGlobalStatus(rows[0]["duplicate_payment_id"] as string);
    return res.json(rows[0]);
  } catch (e) {
    console.error("Update review status error:", e);
    return res.status(500).json({ error: "Failed to update status" });
  }
});

// ── Helper: sync global status ────────────────────────────────────────────────
async function syncGlobalStatus(duplicatePaymentId: string): Promise<void> {
  try {
    const reviews = await query(
      "SELECT status FROM dup_user_reviews WHERE duplicate_payment_id = $1",
      [duplicatePaymentId]
    );
    if (!reviews.length) return;

    const statuses = reviews.map((r) => String(r["status"]));
    const confirmedCount = statuses.filter((s) => s === "confirmed_duplicate").length;
    const dismissedCount = statuses.filter((s) => s === "dismissed").length;
    const underReviewCount = statuses.filter((s) => s === "under_review").length;
    const total = statuses.length;

    let globalStatus: string;
    if (confirmedCount > total / 2) globalStatus = "confirmed_duplicate";
    else if (dismissedCount > total / 2) globalStatus = "dismissed";
    else if (underReviewCount > 0 || confirmedCount > 0) globalStatus = "under_review";
    else globalStatus = "pending";

    await execute("UPDATE dup_duplicate_payments SET status = $1 WHERE id = $2", [globalStatus, duplicatePaymentId]);
  } catch (e) {
    console.error("Sync global status error:", e);
  }
}

export default router;

File: artifacts/py-api/routes/user_reviews.py
"""
User Reviews API — per-analyst duplicate payment review workflow.

Flow:
1. GET  /py-api/user-reviews/users            → list all 10 analysts
2. POST /py-api/user-reviews/fetch            → Master+Text-to-SQL fetch → assign to user
3. GET  /py-api/user-reviews/{user_id}        → list this user's review queue
4. GET  /py-api/user-reviews/{user_id}/{review_id}/opinions → get detector opinions
5. POST /py-api/user-reviews/{user_id}/{review_id}/chat     → chat with Review Agent
6. PATCH /py-api/user-reviews/{user_id}/{review_id}/status  → direct status update
"""
import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from datetime import datetime, timezone

from db import (
    get_db, PREDEFINED_USERS,
    UserRecord, UserReviewRecord, UserReviewMessageRecord,
    DuplicatePaymentRecord,
)
from agents.review_agent import run_review_agent
from agents.detector_agents import get_all_detector_opinions
from agents.master_agent import run_master_agent

router = APIRouter()


# ── helpers ───────────────────────────────────────────────────────────────────

def _sync_global_status(rev: UserReviewRecord, new_status: str, db: Session) -> None:
    """Mirror an analyst's review decision back to the duplicate payments DB.

    Last writer wins — whichever analyst acted most recently sets the global
    status.  The main payments table (PaymentRecord) is never touched.
    """
    dup = db.query(DuplicatePaymentRecord).filter(
        DuplicatePaymentRecord.id == rev.duplicate_payment_id
    ).first()
    if dup:
        dup.status = new_status


def _review_to_dict(rev: UserReviewRecord, dup: DuplicatePaymentRecord) -> dict:
    return {
        "id": rev.id,
        "userId": rev.user_id,
        "duplicatePaymentId": rev.duplicate_payment_id,
        "status": rev.status,
        "notes": rev.notes,
        "createdAt": rev.created_at.isoformat() if rev.created_at else None,
        "updatedAt": rev.updated_at.isoformat() if rev.updated_at else None,
        "payment": {
            "id": dup.id,
            "payment1Id": dup.payment1_id,
            "payment2Id": dup.payment2_id,
            "paymentDate1": dup.payment_date1,
            "paymentDate2": dup.payment_date2,
            "paymentSystem": dup.payment_system,
            "duplicateType": dup.duplicate_type,
            "amount": dup.amount,
            "currency": dup.currency,
            "probability": dup.probability,
            "senderBIC": dup.sender_bic,
            "receiverBIC": dup.receiver_bic,
            "originatorCountry": dup.originator_country,
            "beneficiaryCountry": dup.beneficiary_country,
            "matchedFields": dup.matched_fields or [],
            "status": dup.status,
            "notes": dup.notes,
        },
    }


# ── 1. List all analysts ──────────────────────────────────────────────────────

@router.get("/users")
def list_users():
    return {"users": PREDEFINED_USERS}


# ── 2. Fetch duplicate payments and assign to user ────────────────────────────

@router.post("/fetch")
async def fetch_duplicate_payments(body: dict, db: Session = Depends(get_db)):
    """
    Master Agent + Text-to-SQL query the DB for pending duplicate payments,
    then assign any unreviewed records to this user.
    """
    user_id = body.get("userId", "")
    user = db.query(UserRecord).filter(UserRecord.id == user_id).first()
    if not user:
        raise HTTPException(404, "User not found")

    # Query all duplicate payments (limit to most recent 200 for performance)
    all_dups = (
        db.query(DuplicatePaymentRecord)
        .order_by(DuplicatePaymentRecord.detected_at.desc())
        .limit(200)
        .all()
    )

    # Find which ones this user already has a review for
    existing_ids = {
        r.duplicate_payment_id
        for r in db.query(UserReviewRecord)
        .filter(UserReviewRecord.user_id == user_id)
        .all()
    }

    # Create review records for any not yet assigned
    new_reviews = []
    for dup in all_dups:
        if dup.id not in existing_ids:
            rev = UserReviewRecord(
                id=str(uuid.uuid4()),
                user_id=user_id,
                duplicate_payment_id=dup.id,
                status="pending",
                created_at=datetime.now(timezone.utc),
                updated_at=datetime.now(timezone.utc),
            )
            db.add(rev)
            new_reviews.append(rev)
    db.commit()

    total_assigned = len(existing_ids) + len(new_reviews)

    # Master Agent generates a brief summary
    summary_prompt = (
        f"You have fetched {len(all_dups)} duplicate payment records from the database "
        f"for analyst {user.display_name}. Of these, {len(new_reviews)} are newly assigned "
        f"and {len(existing_ids)} were already in their review queue. "
        f"Please provide a brief 2-sentence summary of what the analyst should expect to review, "
        f"mentioning the total count and suggesting they start with the highest probability records."
    )
    try:
        summary = await run_master_agent(summary_prompt, [], "")
    except Exception:
        summary = (
            f"Fetched {len(all_dups)} duplicate payment records for your review queue. "
            f"{len(new_reviews)} new records have been assigned — start with the highest probability pairs."
        )

    return {
        "totalFetched": len(all_dups),
        "newlyAssigned": len(new_reviews),
        "totalAssigned": total_assigned,
        "summary": summary,
    }


# ── 3. List user's review queue ───────────────────────────────────────────────

@router.get("/{user_id}")
def get_user_reviews(
    user_id: str,
    status: str = "",
    page: int = 1,
    page_size: int = 25,
    db: Session = Depends(get_db),
):
    user = db.query(UserRecord).filter(UserRecord.id == user_id).first()
    if not user:
        raise HTTPException(404, "User not found")

    query = db.query(UserReviewRecord).filter(UserReviewRecord.user_id == user_id)
    if status:
        query = query.filter(UserReviewRecord.status == status)

    total = query.count()
    reviews = (
        query.order_by(UserReviewRecord.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    dup_ids = [r.duplicate_payment_id for r in reviews]
    dup_map = {
        d.id: d
        for d in db.query(DuplicatePaymentRecord)
        .filter(DuplicatePaymentRecord.id.in_(dup_ids))
        .all()
    }

    items = []
    for rev in reviews:
        dup = dup_map.get(rev.duplicate_payment_id)
        if dup:
            items.append(_review_to_dict(rev, dup))

    return {
        "reviews": items,
        "total": total,
        "page": page,
        "pageSize": page_size,
        "totalPages": max(1, (total + page_size - 1) // page_size),
        "user": {"id": user.id, "username": user.username, "displayName": user.display_name},
    }


# ── 4. Get detector opinions for a specific review ────────────────────────────

@router.get("/{user_id}/{review_id}/opinions")
async def get_review_opinions(user_id: str, review_id: str, db: Session = Depends(get_db)):
    rev = db.query(UserReviewRecord).filter(
        UserReviewRecord.id == review_id,
        UserReviewRecord.user_id == user_id,
    ).first()
    if not rev:
        raise HTTPException(404, "Review not found")

    dup = db.query(DuplicatePaymentRecord).filter(
        DuplicatePaymentRecord.id == rev.duplicate_payment_id
    ).first()
    if not dup:
        raise HTTPException(404, "Duplicate payment record not found")

    payment_data = {
        "id": dup.id,
        "payment_system": dup.payment_system,
        "amount": dup.amount,
        "currency": dup.currency,
        "sender_bic": dup.sender_bic,
        "receiver_bic": dup.receiver_bic,
        "payment_date1": dup.payment_date1,
        "payment_date2": dup.payment_date2,
        "matched_fields": dup.matched_fields or [],
        "probability": dup.probability,
        "duplicate_type": dup.duplicate_type,
    }

    result = await get_all_detector_opinions([payment_data])
    return {"opinions": result.get("opinions", []), "consensus": result.get("consensus", {})}


# ── 5. Chat with Review Agent ─────────────────────────────────────────────────

@router.post("/{user_id}/{review_id}/chat")
async def chat_with_review_agent(
    user_id: str, review_id: str, body: dict, db: Session = Depends(get_db)
):
    rev = db.query(UserReviewRecord).filter(
        UserReviewRecord.id == review_id,
        UserReviewRecord.user_id == user_id,
    ).first()
    if not rev:
        raise HTTPException(404, "Review not found")

    user = db.query(UserRecord).filter(UserRecord.id == user_id).first()
    if not user:
        raise HTTPException(404, "User not found")

    dup = db.query(DuplicatePaymentRecord).filter(
        DuplicatePaymentRecord.id == rev.duplicate_payment_id
    ).first()
    if not dup:
        raise HTTPException(404, "Duplicate payment record not found")

    message = body.get("message", "")
    detector_opinions = body.get("detectorOpinions", [])

    payment_data = {
        "id": dup.id,
        "payment1Id": dup.payment1_id,
        "payment2Id": dup.payment2_id,
        "paymentSystem": dup.payment_system,
        "duplicateType": dup.duplicate_type,
        "amount": dup.amount,
        "currency": dup.currency,
        "probability": dup.probability,
        "senderBIC": dup.sender_bic,
        "receiverBIC": dup.receiver_bic,
        "paymentDate1": dup.payment_date1,
        "paymentDate2": dup.payment_date2,
        "matchedFields": dup.matched_fields or [],
        "status": rev.status,
        "_reviewer_name": user.display_name,
    }

    # Restart recovery: load the analyst's FULL message history across ALL their reviews.
    # The Training Agent memory is keyed by user_id, so we need the complete history.
    user_review_ids = [
        r.id
        for r in db.query(UserReviewRecord)
        .filter(UserReviewRecord.user_id == user_id)
        .all()
    ]
    history_records = (
        db.query(UserReviewMessageRecord)
        .filter(UserReviewMessageRecord.review_id.in_(user_review_ids))
        .order_by(UserReviewMessageRecord.timestamp)
        .all()
    )
    db_history = [{"role": m.role, "content": m.content} for m in history_records]

    result = await run_review_agent(
        user_message=message,
        user_id=user_id,
        reviewer_name=user.display_name,
        payment_data=payment_data,
        detector_opinions=detector_opinions,
        db_history=db_history,
    )

    # Persist messages
    db.add(UserReviewMessageRecord(
        id=str(uuid.uuid4()),
        review_id=review_id,
        role="user",
        content=message,
        timestamp=datetime.now(timezone.utc),
    ))
    db.add(UserReviewMessageRecord(
        id=str(uuid.uuid4()),
        review_id=review_id,
        role="assistant",
        content=result["response"],
        timestamp=datetime.now(timezone.utc),
    ))

    # Apply status update if agent triggered one
    if result.get("statusUpdate"):
        rev.status = result["statusUpdate"]
        rev.updated_at = datetime.now(timezone.utc)
        _sync_global_status(rev, result["statusUpdate"], db)

    db.commit()

    return {
        "response": result["response"],
        "statusUpdate": result.get("statusUpdate"),
        "currentStatus": rev.status,
    }


# ── 6. Get conversation history for a review ─────────────────────────────────

@router.get("/{user_id}/{review_id}/messages")
def get_review_messages(user_id: str, review_id: str, db: Session = Depends(get_db)):
    rev = db.query(UserReviewRecord).filter(
        UserReviewRecord.id == review_id,
        UserReviewRecord.user_id == user_id,
    ).first()
    if not rev:
        raise HTTPException(404, "Review not found")

    messages = (
        db.query(UserReviewMessageRecord)
        .filter(UserReviewMessageRecord.review_id == review_id)
        .order_by(UserReviewMessageRecord.timestamp)
        .all()
    )

    return {
        "messages": [
            {"role": m.role, "content": m.content,
             "timestamp": m.timestamp.isoformat() if m.timestamp else None}
            for m in messages
        ],
        "currentStatus": rev.status,
    }


# ── 7. Direct status update (without agent) ───────────────────────────────────

@router.patch("/{user_id}/{review_id}/status")
def update_review_status(
    user_id: str, review_id: str, body: dict, db: Session = Depends(get_db)
):
    rev = db.query(UserReviewRecord).filter(
        UserReviewRecord.id == review_id,
        UserReviewRecord.user_id == user_id,
    ).first()
    if not rev:
        raise HTTPException(404, "Review not found")

    new_status = body.get("status", "")
    valid = {"pending", "confirmed_duplicate", "dismissed", "under_review"}
    if new_status not in valid:
        raise HTTPException(400, f"Invalid status. Must be one of: {valid}")

    rev.status = new_status
    rev.notes = body.get("notes", rev.notes)
    rev.updated_at = datetime.now(timezone.utc)
    _sync_global_status(rev, new_status, db)
    db.commit()

    return {"success": True, "status": rev.status, "reviewId": rev.id}
